business/openid_auth.go

```diff
@@ -28,8 +38,141 @@ type OpenIdMetadata struct {
 	ResponseTypesSupported []string `json:"response_types_supported"`
 }
 
+type OpenIdCallbackParams struct {
+	Code          string
+	ExpiresOn     time.Time
+	IdToken       string
+	Nonce         string
+	NonceHash     []byte
+	ParsedIdToken *jwt.Token
+	State         string
+	Subject       string
+}
+
 var cachedOpenIdMetadata *OpenIdMetadata
 
+func BuildOpenIdJwtClaims(openIdParams *OpenIdCallbackParams) *config.IanaClaims {
+	return &config.IanaClaims{
+		SessionId: openIdParams.IdToken,
+		StandardClaims: jwt.StandardClaims{
+			Subject:   openIdParams.Subject,
+			ExpiresAt: openIdParams.ExpiresOn.Unix(),
+			Issuer:    config.AuthStrategyOpenIdIssuer,
+		},
+	}
+}
+
+func ExtractOpenIdCallbackParams(w http.ResponseWriter, r *http.Request) (params *OpenIdCallbackParams, err error) {
+	params = &OpenIdCallbackParams{}
+
+	// Get the nonce code hash
+	var nonceCookie *http.Cookie
+	if nonceCookie, err = r.Cookie(OpenIdNonceCookieName); err == nil {
+		params.Nonce = nonceCookie.Value
+
+		hash := sha256.Sum224([]byte(nonceCookie.Value))
+		params.NonceHash = make([]byte, sha256.Size224)
+		copy(params.NonceHash, hash[:])
+	}
+
+	// Parse/fetch received form data
+	err = r.ParseForm()
+	if err != nil {
+		err = fmt.Errorf("error parsing form info: %w", err)
+	} else {
+		// Read relevant form data parameters
+		params.Code = r.Form.Get("code")
+		params.IdToken = r.Form.Get("id_token")
+		params.State = r.Form.Get("state")
```

Comment:
Are these all form parms necessary/required? If so, should we be checking to see if Code, IdToken, and/or State are `nil` and if so generate an error here? I am wondering if we should be performing some sanity checks here in case we hit some odd OpenID implementation that doesn't provide one or more of these.

Reply:
HA!  I just noticed in code further below (`CheckOpenIdAuthenticationCodeFlowParams`) it does check Code and State. But what about IdToken?

Reply:
IdToken is for the other flow. Not used in this PR, but it will be used in the follow-up PR.